cmake/Modules/FindDPDK.cmake

```diff
@@ -0,0 +1,181 @@
+#  DPDK_LIBRARIES, the libraries needed by a DPDK user
+#  DPDK_CFLAGS_OTHER, the compile flags to use
+#  DPDK_VERSION, the version of the library
+# ~~~
+
+include(FindPackageHandleStandardArgs)
+
+function(DPDK_READ_VERSION DPDK_VERSION DPDK_VERSION_FILE)
+  if(NOT DPDK_VERSION_FILE)
+    return()
+  endif()
+
+  file(READ "${DPDK_VERSION_FILE}" DPDK_VERSION_STR)
+  string(
+    REGEX MATCH
+          "#define RTE_VER_YEAR ([0-9]+)"
+          _
+          ${DPDK_VERSION_STR})
+  set(DPDK_VERSION_MAJOR ${CMAKE_MATCH_1})
+
+  string(
+    REGEX MATCH
+          "#define RTE_VER_MONTH ([0-9]+)"
+          _
+          ${DPDK_VERSION_STR})
+  set(DPDK_VERSION_MINOR ${CMAKE_MATCH_1})
+
+  string(
+    REGEX MATCH
+          "#define RTE_VER_MINOR ([0-9]+)"
+          _
+          ${DPDK_VERSION_STR})
+  set(DPDK_VERSION_PATCH ${CMAKE_MATCH_1})
+
+  set(DPDK_VERSION
+      "${DPDK_VERSION_MAJOR}.${DPDK_VERSION_MINOR}.${DPDK_VERSION_PATCH}"
+      PARENT_SCOPE)
+endfunction()
+
+# Try to find DPDK with pkg-config first!
+find_package(PkgConfig QUIET)
+if(PKG_CONFIG_FOUND)
+  pkg_check_modules(
+    DPDK
+    QUIET
+    IMPORTED_TARGET
+    libdpdk>=18.11)
+endif()
+
+# We found using Pkg-Config!
+if(DPDK_FOUND)
+  message("DPDK found with pkg-config")
+  if(DPDK_DEBUG)
+    message("-----------")
+    message("Libraries: ${DPDK_LIBRARIES}")
+    message("Link Libraries: ${DPDK_LINK_LIBRARIES}")
+    message("Library DIr: ${DPDK_LIBRARY_DIRS}")
+    message("Ldflags: ${DPDK_LDFLAGS}")
+    message("Include Dirs: ${DPDK_INCLUDE_DIRS}")
+    message("Cflags: ${DPDK_CFLAGS}")
+    message("Cflags Other: ${DPDK_CFLAGS_OTHER}")
+    message("Version: ${DPDK_VERSION}")
+    message("-----------")
+  endif()
+else()
+  message("DPDK not found with pkg-config trying legacy mode")
+
+  # Find the include dirs and get the version
+  find_path(DPDK_VERSION_INCLUDE_DIR rte_version.h REQUIRED PATH_SUFFIXES dpdk include)
+  find_path(DPDK_CONFIG_INCLUDE_DIR rte_config.h REQUIRED PATH_SUFFIXES dpdk include)
+
+  set(DPDK_INCLUDE_DIRS ${DPDK_INCLUDE_DIRS})
+  list(APPEND DPDK_INCLUDE_DIRS ${DPDK_VERSION_INCLUDE_DIR})
+  list(APPEND DPDK_INCLUDE_DIRS ${DPDK_CONFIG_INCLUDE_DIR})
+  list(REMOVE_DUPLICATES DPDK_INCLUDE_DIRS)
+
+  dpdk_read_version(DPDK_VERSION "${DPDK_VERSION_INCLUDE_DIR}/rte_version.h")
+
+  # Get all the libraries regarding the version
+  list(
+    APPEND
+    _DPDK_LOOK_FOR_LIBS
+    net
+    kni
+    ethdev
+    mbuf
+    eal
+    mempool
+    ring
+    kvargs
+    hash
+    cmdline
+    pci
+    bus_pci
+    bus_vdev
+    mempool_ring)
+
+  if(DPDK_VERSION VERSION_LESS "20.11")
+    list(
+      APPEND
+      _DPDK_LOOK_FOR_LIBS
+      pmd_bond
+      pmd_vmxnet3_uio
+      pmd_virtio
+      pmd_enic
+      pmd_i40e
+      pmd_fm10k
+      pmd_ixgbe
+      pmd_e1000
+      pmd_ring
+      pmd_af_packet)
+  endif()
+
+  # Check that all libraries exists
+  foreach(lib ${_DPDK_LOOK_FOR_LIBS})
+    find_library(rte_${lib} NAMES rte_${lib} REQUIRED)
```

Comment:
~~~cmake
find_library(rte_${lib} NAMES rte_${lib} rte_${lib}_uio NAMES_PER_DIR REQUIRED)
~~~
to generally catch an `_uio` suffix.

(You should never get `rte_pmd_vmxnet3_uio-NOTFOUND` due to the `REQUIRED` ... but this keyword requires CMake 3.18. If you want to support older versions of CMake, you must explicitly catch NOTFOUND.)

Reply:
Yes but Cmake 3.18 is a bit to recent. 

I will catch the Not found and add a specific fallback

Reply:
@dg0yt seems that your first solution is ok with cmake 3.12 no ? https://cmake.org/cmake/help/v3.12/command/find_library.html

Reply:
My solution is sufficient to deal with the _uio suffix, but it is not sufficent for the general case of not-found.